exporter/elasticsearchexporter/pdata_serializer.go

```diff
@@ -0,0 +1,372 @@
+	writeScope(v, scope, scopeSchemaURL, false)
+	writeLogBody(v, record)
+	_ = v.OnObjectFinished()
+	return buf.Bytes(), nil
+}
+
+func writeDataStream(v *json.Visitor, attributes pcommon.Map) {
+	_ = v.OnKey("data_stream")
+	_ = v.OnObjectStart(-1, structform.AnyType)
+	attributes.Range(func(k string, val pcommon.Value) bool {
+		if strings.HasPrefix(k, "data_stream.") && val.Type() == pcommon.ValueTypeStr {
+			writeStringFieldSkipDefault(v, k[12:], val.Str())
+		}
+		return true
+	})
+
+	_ = v.OnObjectFinished()
+}
+
+func writeLogBody(v *json.Visitor, record plog.LogRecord) {
+	if record.Body().Type() == pcommon.ValueTypeEmpty {
+		return
+	}
+	_ = v.OnKey("body")
+	_ = v.OnObjectStart(-1, structform.AnyType)
+
+	// Determine if this log record is an event, as they are mapped differently
+	// https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/events.md
+	var bodyType string
+	if _, hasEventNameAttribute := record.Attributes().Get("event.name"); hasEventNameAttribute || record.EventName() != "" {
+		bodyType = "structured"
+	} else {
+		bodyType = "flattened"
+	}
+	body := record.Body()
+	switch body.Type() {
+	case pcommon.ValueTypeMap:
+	case pcommon.ValueTypeSlice:
+		// output must be an array of objects due to ES limitations
+		// otherwise, wrap the array in an object
+		s := body.Slice()
+		allMaps := true
+		for i := 0; i < s.Len(); i++ {
+			if s.At(i).Type() != pcommon.ValueTypeMap {
+				allMaps = false
+			}
+		}
+
+		if !allMaps {
+			body = pcommon.NewValueMap()
+			m := body.SetEmptyMap()
+			record.Body().Slice().CopyTo(m.PutEmptySlice("value"))
+		}
+	default:
+		bodyType = "text"
+	}
+	_ = v.OnKey(bodyType)
+	writeValue(v, body, false)
+	_ = v.OnObjectFinished()
+}
+
+func writeResource(v *json.Visitor, resource pcommon.Resource, resourceSchemaURL string, stringifyMapAttributes bool) {
+	_ = v.OnKey("resource")
+	_ = v.OnObjectStart(-1, structform.AnyType)
+	writeStringFieldSkipDefault(v, "schema_url", resourceSchemaURL)
+	writeAttributes(v, resource.Attributes(), stringifyMapAttributes)
+	writeIntFieldSkipDefault(v, "dropped_attributes_count", int64(resource.DroppedAttributesCount()))
+	_ = v.OnObjectFinished()
+}
+
+func writeScope(v *json.Visitor, scope pcommon.InstrumentationScope, scopeSchemaURL string, stringifyMapAttributes bool) {
+	_ = v.OnKey("scope")
+	_ = v.OnObjectStart(-1, structform.AnyType)
+	writeStringFieldSkipDefault(v, "schema_url", scopeSchemaURL)
+	writeStringFieldSkipDefault(v, "name", scope.Name())
+	writeStringFieldSkipDefault(v, "version", scope.Version())
+	writeAttributes(v, scope.Attributes(), stringifyMapAttributes)
+	writeIntFieldSkipDefault(v, "dropped_attributes_count", int64(scope.DroppedAttributesCount()))
+	_ = v.OnObjectFinished()
+}
+
+func writeAttributes(v *json.Visitor, attributes pcommon.Map, stringifyMapValues bool) {
+	if attributes.Len() == 0 {
+		return
+	}
+	attrCopy := pcommon.NewMap()
+	attributes.CopyTo(attrCopy)
```

Comment:
It would be great if we could avoid this copy. The `geo.location` merge might be a bit complicated, but the data stream fields should be reasonably straightforward to filter out by iterating through. A job for another day.

Reply:
Done in 539fa9d

Reply:
Relatively speaking, it didn’t reduce allocations a lot. I think the biggest gain would be pooling the buffers. Does docappender support that already?

Reply:
There's a buffer per bulkIndexer, but there is no `sync.Pool`.